### PR TITLE
KAFKA-3995: fix compression ratio estimation.

### PR DESCRIPTION
**Pull request by becketqin**

### CODE REVIEW

**Comment by becketqin** [comment]
@jjkoshy Do you have time to take a look? Thanks.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadata.java
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadata.java]: Couple of comments:

- Although you don't seem to need to in the tests, wouldn't it be better to allow using our `Time` interface so that it can be mocked if necessary?
- Is this necessary in the first place? i.e., if there is a `nextRecordMetadata` shouldn't the preceding ones in the chain be done in the first place? i.e., if we are here then there was a split and therefore those earlier results should be available, so zero time budget will go toward checking those.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadata.java
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadata.java]: overkill?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:122]: It would help to add some detail to the method name to highlight this.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: Rather than reset estimation in `Sender`, would it be possible to do that here? Reason being - the actual splitting isn't explicitly done. It is done (IIUC) by virtue of the fact that the estimation has been reset. So it would be clearer if that logic is fully contained in this method. I haven't looked closely enough to verify that this would still work, but the point remains that the `split` is effective only if the estimation has been reset (again, IIUC).

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java]: See comment in `ProducerBatch.split`

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: I understand the name, but would `createBatchOffAccumulator` be a better name?

Comment (jjkoshy) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: Separately though, should we actually do this off-accumulator instead of on-accumulator? Are you concerned about a resource deadlock if the accumulator is full and every batch needs to be split?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java
Comment (jjkoshy) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: Also, is there a clearer name for the `batchSize` argument? e.g., `splitBatchSize` or `targetBatchSize`?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java]: I may have missed this, but for the split batches I don't see any special handling for _not_ deallocating them. i.e., since those were created off-the-book, shouldn't their deallocation not be accounted in the accumulator? The off-accumulator batches should not be returned to the accumulator.

File: clients/src/main/java/org/apache/kafka/common/record/AbstractRecords.java
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/common/record/AbstractRecords.java]: Do we really need this additional `ByteBuffer`-specific method vs just passing in its underlying `array` to the preceding method?

File: clients/src/main/java/org/apache/kafka/common/record/CompressionRatioEstimator.java
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/common/record/CompressionRatioEstimator.java]: Do we really need an explicit initialization vs optimistic initialization from `updateEstimation`?

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java]: Similar comment as in `AbstractRecords`

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: Similar comment as in `AbstractRecords`

File: clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java]: I find it slightly weird to update the ratio here. Yes it is the last operation, but clearer to do it outside (in the `close` method itself) no?

File: clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java]: Consider renaming this to `actualCompressionRatio` just to enhance the distinction from `estimatedCompressionRatio`.

File: clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java:672]: Pre-existing: in estimating whether there is room, we don't apply the estimated compression to the inbound record but assume a compression ratio of 1.0. This is conservative, but may be overly conservative. Do you know why this was done? Since it is not obvious and if you know why it would be worth adding a comment (line 670)

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:490]: Just to confirm: split batches get the full retry budget right? How about producer request timeouts?

Comment (jjkoshy) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:490]: If there are nuances to these, then it should be mentioned in the producer's javadoc

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java]: Isn't logging this redundant to including `RecordBatchTooLargeException` earlier in the message?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java]: Also, I'm a little curious about the order you chose here. Why deallocate after the split and not before?

**Review by jjkoshy — CHANGES_REQUESTED** [review]
Thank you for the patch. Here are my initial comments.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadata.java
Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadata.java]: This is to address the case that users call something like `get(Long.MAX_VALUE, TimeUnit.MILLISECONDS)`, which seems still possible even users should call `get()` without any argument in that case.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadata.java
Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadata.java]: The timeout budget was passed by the users, for example, if users call `get(1, TimeUnit.SECONDS)` that means the user wants to get a result within one second. It is possible that say after 100 milliseconds, the batch is split and resent. The first result is then waken up and now we need to wait for the new results of the batches after the split. And there are only 900 milliseconds left before we should return the result, whether succeed or fail.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java
Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: It is essentially off the buffer pool.

The main concern here is that the split takes place in the sender thread, so if we allocate on-accumulator and block on buffer full will result in deadlock. If we ignore the buffer pool, it is essentially off-accumulator. We can potentially allocate buffer on-accumulator forcefully, but there are a few downside:
1. The buffer pool semantic essentially becomes weaker.
2. The free blocks allocated forcefully should not be returned to the accumulator otherwise the accumulator may take more memory than it actually needed. 
3. Also if we forcefully allocate memory from the buffer pool, the available memory size reported by the pool may be negative, which is also a little weird.

Personally I think it is simpler to just allocate the memory outside the buffer pool given the split should be rare.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java
Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:490]: Could elaborate a little bit on the question? The split batches will be treated as if they were never sent out before. So there is nothing special about the them.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java
Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java]: Good catch. We should not return the split batches to the accumulator.

Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java]: The deallocation is for the original big batch. Since we are still holding the original batch when splitting, we don't want to deallocate it. That is why we deallocate the batch after splitting.

File: clients/src/main/java/org/apache/kafka/common/record/AbstractRecords.java
Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/common/record/AbstractRecords.java]: The reason we need it is that when splitting the batch, we have a big array holding all the messages and we get a ByteBuffer sliced from it. So there is no dedicated underlying array. And we don't want to do a memory copy to get the message size in this case. An alternative is to expose some size information from the `MemoryRecords` which is a little hacky.

File: clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java
Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java]: ack

File: clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java
Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java]: ack

File: clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java
Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java:672]: ack

File: clients/src/main/java/org/apache/kafka/common/record/CompressionRatioEstimator.java
Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/common/record/CompressionRatioEstimator.java]: ack

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java
Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java]: ack

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java
Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: ack

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java
Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:122]: ack

**Review by becketqin — COMMENTED** [review]
@jjkoshy Thanks for the review. I answered some of the questions and want to understand a bit more on some others. Will update the patch shortly.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadata.java
Comment (jjkoshy) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadata.java]: Got it.

File: clients/src/main/java/org/apache/kafka/common/record/AbstractRecords.java
Comment (jjkoshy) [reply clients/src/main/java/org/apache/kafka/common/record/AbstractRecords.java]: Thanks for clarifying the fact that these are slices. As discussed offline let us see if we can consolidate the logic (i.e., wrap the `byte[]` objects in a `ByteBuffer` and delegate to the other method).

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadata.java
Comment (jjkoshy) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadata.java]: Sounds good. Minor: may want to switch to `Long.MAX_VALUE` on overflow for `deadline`.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java
Comment (jjkoshy) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: Ok cool - should we have a separate sensor for overallocated memory?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java
Comment (jjkoshy) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:490]: Discussed offline

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java
Comment (jjkoshy) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java]: Thanks - will look in a bit.

Comment (jjkoshy) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java]: Cool - can you add a comment explaining this?

File: clients/src/main/java/org/apache/kafka/common/record/CompressionRatioEstimator.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/CompressionRatioEstimator.java:32]: Hmm, what's the reasoning for making this a JVM singleton?

Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/common/record/CompressionRatioEstimator.java:32]: I wondered that too. The compression ratio has always been a JVM singleton. I guess it kind of make sense if think of the compression ratio as topic specific. i.e. the compression ratio estimation is globally applicable to a topic in the same JVM.

Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/record/CompressionRatioEstimator.java:32]: I hadn't noticed that, but it's true that it was a JVM singleton. And updated without any locking. Have you had a chance to run some benchmarks on this branch to figure out if there's any impact in the case where no splitting happens?

**Comment by becketqin** [comment]
I realized that with KIP-98 change, we need to ensure sequence id of the batches are also updated correctly. I'll update the patch to fix that.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java
Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: Because the split happens rarely, it seems that this sensor is going to be 0 most of the times. If there are frequent split, we will see that from the split rate sensor. So I am not sure if the sensor is needed. Also this sensor is not in the original KIP. Maybe we can add it later if it turns out to be useful.

File: clients/src/main/java/org/apache/kafka/common/record/CompressionRatioEstimator.java
Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/common/record/CompressionRatioEstimator.java:32]: I haven't done benchmark. Is the concern about the new synchronization? It is probably not a big issue given that there will only be one read and one write to this map per batch. A single producer benchmark is unlikely to give any insight in this case. We will need many producers to produce in the same JVM at the same time to make the synchronization take some effect if any. So personally I am not really concerned about this synchronization.

Comment (becketqin) [reply clients/src/main/java/org/apache/kafka/common/record/CompressionRatioEstimator.java:32]: Sorry I probably misunderstood your concern. Is it because of the new way to update compression ratio? I just ran producer-performance benchmark, the new way provides better and more stable throughput than the old one in general, which is expected as the new way should be less sensitive to the compression ratio change of a single batch.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java
Comment (jjkoshy) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java]: Nit: would generally prefer to see this made explicit (i.e., `drain` twice) as opposed to loop (which in case of some bug could in the worst case be non-terminating).

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java
Comment (jjkoshy) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java]: Remove

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java
Comment (jjkoshy) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java]: similar comment here

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java]: Rather than this, would it be simpler to just have a boolean on the batch indicating whether it is off-accumulator or not?

**Review by jjkoshy — COMMENTED** [review]
+1 although can you please respond to/address the remaining minor comments?

**Comment by becketqin** [comment]
@jjkoshy Thanks a lot for the review! Just updated the patch addressing your comments. Will merge the patch after the tests passes.

**Comment by hachikuji** [comment]
@becketqin Thanks for the patch. I looked over the sequence number update logic on splitting and it seems correct. Are you planning to cherry-pick this into 0.11.0?

**Comment by becketqin** [comment]
@hachikuji Thanks for the review. Yes, I prefer to have this in 0.11.0 if possible. But I'll let @ijuma make the call. @ijuma what do you think of cherry-picking this patch to 0.11.0?

**Comment by ijuma** [comment]
@becketqin, if you, @jjkoshy and Jason think this is a low risk change, I'm OK with cherry-picking it. It's a little larger than I would like at this stage, but keeping trunk and 0.11.0 similar will probably help with some of the in-flight items.

**Comment by becketqin** [comment]
@ijuma I spoke to Joel and he is neutral on this. I personally prefer to have it in 0.11.0 but given it is a long existing issue, waiting for another release is also fine with me.

@hachikuji What do you think?

**Comment by hachikuji** [comment]
I'm fine with cherry-picking. It's useful to keep 0.11.0 from diverging too far for the time being.

**Comment by becketqin** [comment]
Cool it seems there isn't too much concern about cherry-picking the patch to 0.11.0. I'll just do it then.

**Pull request closed without merging**
